File: src/main/java/im/joker/config/filter/AuthFilter.java
package im.joker.config.filter;

import im.joker.device.IDevice;
import im.joker.exception.ErrorCode;
import im.joker.exception.ImException;
import im.joker.helper.GlobalStateHolder;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.reactivestreams.Publisher;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.stereotype.Component;
import org.springframework.util.CollectionUtils;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebFilter;
import org.springframework.web.server.WebFilterChain;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.util.function.Tuple2;
import reactor.util.function.Tuples;

import java.time.Duration;
import java.util.List;

import static im.joker.exception.ErrorCode.FORBIDDEN;

@Component
@Slf4j
public class AuthFilter implements WebFilter {

    private static final String LOGIN_DEVICE = "login_device";

    @Autowired
    private GlobalStateHolder globalStateHolder;

    private final List<Tuple2<HttpMethod, String>> withoutAuthUrls = List.of(
            Tuples.of(HttpMethod.POST, "/_matrix/client/r0/register"),
            Tuples.of(HttpMethod.GET, "/_matrix/client/r0/login"),
            Tuples.of(HttpMethod.GET, "/_matrix/client/versions"),
            Tuples.of(HttpMethod.POST, "/_matrix/client/r0/login")
    );

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {

        ServerHttpRequest request = exchange.getRequest();
        String url = request.getPath().value();
        HttpMethod method = request.getMethod();
        boolean exists = withoutAuthUrls.stream().anyMatch(e -> e.getT1().equals(method) && e.getT2().equals(url));
        if (HttpMethod.OPTIONS.equals(method) || exists) {
            return chain.filter(exchange);
        }
        List<String> headers = request.getHeaders().get(HttpHeaders.AUTHORIZATION);
        String token = exchange.getAttribute("access_token");
        if (CollectionUtils.isEmpty(headers) && StringUtils.isBlank(token)) {
            throw new ImException(ErrorCode.UNAUTHORIZED, HttpStatus.UNAUTHORIZED);
        }
        if (!CollectionUtils.isEmpty(headers)) {
            token = StringUtils.substringAfter(headers.get(0), "Bearer ");
        }
        if (StringUtils.equals(token, "undefined")) {
            throw new ImException(ErrorCode.UNKNOWN_TOKEN, HttpStatus.FORBIDDEN);
        }
        Mono<IDevice> device = globalStateHolder.getDeviceManager().find(token);
        return device.switchIfEmpty(Mono.error(new ImException(FORBIDDEN, HttpStatus.FORBIDDEN)))
                .flatMap(e -> chain.filter(exchange).subscriberContext(context -> context.put(getLoginDevice(), e)));


    }

    public static String getLoginDevice() {
        return LOGIN_DEVICE;
    }



}

File: src/main/kotlin/im/joker/config/AuthFilter.kt
package im.joker.config

import im.joker.device.Device
import im.joker.exception.ErrorCode
import im.joker.exception.ImException
import im.joker.helper.GlobalStateHolder
import org.apache.commons.lang3.StringUtils
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.util.AntPathMatcher
import org.springframework.util.CollectionUtils
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono
import reactor.util.context.Context
import reactor.util.function.Tuple2
import reactor.util.function.Tuples

/**
 * @Author: mkCen
 * @Date: 2020/11/22
 * @Time: 0:54
 * @Desc:
 */
@Component
class AuthFilter : WebFilter {
    private val log: Logger = LoggerFactory.getLogger(AuthFilter::class.java)

    @Autowired
    lateinit var globalStateHolder: GlobalStateHolder

    companion object {
        const val LOGIN_DEVICE = "login_device"
    }

    private val withoutAuthUrls = listOf(
            Tuples.of(HttpMethod.POST, "/_matrix/client/r0/register"),
            Tuples.of(HttpMethod.GET, "/_matrix/client/r0/login"),
            Tuples.of(HttpMethod.GET, "/_matrix/client/versions"),
            Tuples.of(HttpMethod.POST, "/_matrix/client/r0/login"),
            Tuples.of(HttpMethod.GET, "/_matrix/client/versions/test"),
            Tuples.of(HttpMethod.GET, "/favicon.ico"),
            Tuples.of(HttpMethod.GET, "/_matrix/media/r0/download/**"),
            Tuples.of(HttpMethod.GET, "/_matrix/media/r0/config")
    )

    private val antMatcher = AntPathMatcher()

    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val request = exchange.request
        val url = request.path.value()
        val method = request.method
        val exists = withoutAuthUrls.stream().anyMatch { e: Tuple2<HttpMethod, String> -> e.t1 == method && antMatcher.match(e.t2, url) }
        if (HttpMethod.OPTIONS == method || exists) {
            return chain.filter(exchange)
        }
        val headers = request.headers[HttpHeaders.AUTHORIZATION]
        var token = exchange.getAttribute<String>("access_token")
        headers?.let {
            token = it[0].substringAfter("Bearer ")
        }
        if (token.isNullOrBlank()) {
            throw ImException(ErrorCode.UNAUTHORIZED, HttpStatus.UNAUTHORIZED)
        }

        if (StringUtils.equals(token, "undefined")) {
            throw ImException(ErrorCode.UNKNOWN_TOKEN, HttpStatus.FORBIDDEN)
        }
        val device: Mono<Device> = globalStateHolder.deviceManager.find(token!!)
        return device.switchIfEmpty(Mono.error(ImException(ErrorCode.FORBIDDEN, HttpStatus.FORBIDDEN)))
                .flatMap { e: Device -> chain.filter(exchange).contextWrite { context: Context -> context.put(LOGIN_DEVICE, e) } }

    }


}